Acquire proper locks when releasing resource queue locks from lock table

jimmyyih opened the pull request:
When process tries to release the resource queue lock after deadlock
error, the process was not properly holding the partitionLock and
ResQueueLock. This could lead to shared memory corruption (when
another process goes to release its resource queue locks) and later
cause segmentation faults. Fix the issue by making sure we hold both
locks before touching the shared memory queue. We also introduce a
validation check to PANIC when corruption of the lock table is
encountered.

This commit contains code inspired from Github user dh-cloud's PR:
https://github.com/greenplum-db/gpdb/pull/12185

This commit also contains test code inspired from Github user
dreamedcheng's PR:
https://github.com/greenplum-db/gpdb/pull/9652

Co-authored-by: dh-cloud <60729713+dh-cloud@users.noreply.github.com>
Co-authored-by: wuchengwen <wcw190496@alibaba-inc.com>
Co-authored-by: Yao Wang <wayao@vmware.com>
Co-authored-by: Soumyadeep Chakraborty <soumyadeep2007@gmail.com>
Co-authored-by: Ashwin Agrawal <aashwin@vmware.com>
Co-authored-by: Kate Dontsova <edontsova@pivotal.io>

jimmyyih commented on src/backend/utils/resscheduler/resqueue.c:
```diff
@@ -980,10 +981,24 @@ ResCleanUpLock(LOCK *lock, PROCLOCK *proclock, uint32 hashcode, bool wakeupNeede
+			SHM_QUEUE  *prevElem = (SHM_QUEUE *) MAKE_PTR((queue)->prev);
+			SHMQueueDelete(queue);
+			if (prevElem->next == INVALID_OFFSET || nextElem->prev == INVALID_OFFSET)
+				elog(PANIC, "corruption of lock table encountered");
```
@yaowangm Do you think this PANIC should be under a GUC to do `PANIC ? WARNING`?  Check with other developers to see what they think as well.

asimrp replied:
Invariant for SHMQueue is - next and prev pointers are valid for every member of the queue.  Therefore, the PANIC is a good idea, no GUC is needed.  It will help diagnose such issues faster.  Besides, new GUCs are prohibited in stable branches, if I remember correctly.

zhangh43 replied:
+1 on Asim's point.

yaowangm replied:
@asimrp @zhangh43 
Jimmy and I talked about the behavior this morning, and I also talked with Hubert just now.

We have two options when the lock table is detected to be damaged:

1. Panic immediately and make core dump.
2. Show a warning and continue.

For option 2, customer would have a chance to manually disconnect all clients and shut down the db gracefully to avoid crash recovery which is unavoidable for option 1. So option 2 also makes sense technically. Option 1 should be the default behavior. When customer hit the panic, they could change to option 2 via a GUC and tolerate the error a bit more next time.

However, option 2 has shortcomings:

1. As we known, the rate of the error is very low. So customer is not expected to hit the error again in general.
2. We cannot guarantee that GPDB can continue safely till the graceful shut down. The damaged lock table may cause segment fault whenever other processes try to release lock.

Considering the emergency of the issue, I prefer to deliver the code as-is. We can keep the idea of GUC in mind if it's necessary in future.

asimrp replied:
I only humanity paid attention to warnings, our planet would have been a much better place today!

yaowangm replied:
Another comment...yesterday I added the code of checking prev/next into SHMQueueDelete(), and Deep moved it into ResCleanUpLock() because of performance consideration. It is his comment:

_This is in a super-hot code path. Probably the most serious path is the chain: LockTuple->LockAcquire which implies per-tuple overhead._

It makes sense for me, but the side effect is that we can only monitor the codepath of ResCleanUpLock(), and missed other codepaths calling SHMQueueDelete(). Could we use some kind of wrap function such as SHMQueueDeleteWithCheck() for only vulnerable callings from resource queue to avoid the performance impact?

I am not sure if it matters much because we don't find issue on other codepaths by now. Just some thoughts for your reference.

jimmyyih replied:
Discussed with Ashwin and Deep; we think it's okay to just leave it as PANIC only and avoid having the GUC.

> Could we use some kind of wrap function such as SHMQueueDeleteWithCheck() for only vulnerable callings from resource queue to avoid the performance impact?

I think this could be useful for GPDB 7+.  Maybe we can think about doing that on the master branch if other developers think it would be useful too.

ashwinstar replied:
I am hinting towards if possibility exist to delete all the resource queue code for master/main branch and only focus on resource groups going forward :-) That way the need for this will disappear.

asimrp commented on src/test/isolation2/sql/resource_queue_deadlock.sql:
```diff
@@ -0,0 +1,30 @@
+
+1: SET role role_deadlock_test;
+1: CREATE TABLE t_deadlock_test(c1 int);
+1&: INSERT INTO t_deadlock_test VALUES (1);
```
This should be blocked by the `before_auto_stats` fault.  If that's correct, it will be helpful to mention it in a comment.

zhangh43 replied:
when a query finishes, resource queue will call `ResLockRelease` to release the limit's `current_value`

jimmyyih replied:
Added more comments/details to the new isolation2 test.

asimrp commented on src/test/isolation2/sql/resource_queue_deadlock.sql:
```diff
@@ -0,0 +1,30 @@
+1&: INSERT INTO t_deadlock_test VALUES (1);
+2: SET role role_deadlock_test;
+2: BEGIN;
+2: analyze t_deadlock_test;
```
With my limited understanding of resource queues, this command should be blocked because `active_statemets` limit is already reached due to session 1.  What am I missing?

zhangh43 replied:
when a query finishes, resource queue will call ResLockRelease to release the limit's current_value

jimmyyih replied:
Only certain queries are handled through resource queues.  ANALYZE is not one of them.

From our Greenplum documentation:
>Not all SQL statements submitted through a resource queue are evaluated against the queue limits. By default only SELECT, SELECT INTO, CREATE TABLE AS SELECT, and DECLARE CURSOR statements are evaluated. If the server configuration parameter resource_select_only is set to off, then INSERT, UPDATE, and DELETE statements will be evaluated as well.
>
>Also, an SQL statement that is run during the execution of an EXPLAIN ANALYZE command is excluded from resource queues.

asimrp replied:
I see, that is helpful, thank you!

asimrp commented on src/test/regress/sql/resource_queue_lock_acquire.sql:
```diff
@@ -0,0 +1,21 @@
+CREATE ROLE resource_queue_lock_acquire_role RESOURCE QUEUE foo_resqueue;
+
+-- Set error fault for ResLockAcquire STATUS_OK
+SELECT gp_inject_fault_new('resource_queue_lock_acquire_ok', 'error', dbid) FROM gp_segment_configuration WHERE ROLE = 'p' AND status = 'u' AND content = -1;;
```
It may be only me, but couldn't find a definition for this fault.

jimmyyih replied:
Yeah sorry, this test shouldn't have been added to the PR.  It was part of our attempt to put a fault injector in the `ResLockAcquire` code to simulate a desired error to trigger the try/catch in `ResLockPortal` in a way that we would hit the assertion failure in `ResCleanUpLock`... but that didn't work very well so we adopted the isolation2 test that properly creates a deadlock to go into the `ResCleanUpLock` code path.

ashwinstar commented on src/backend/utils/resscheduler/resqueue.c:
```diff
@@ -980,10 +981,24 @@ ResCleanUpLock(LOCK *lock, PROCLOCK *proclock, uint32 hashcode, bool wakeupNeede
 
 		if (proclock->lockLink.next != INVALID_OFFSET)
-			SHMQueueDelete(&proclock->lockLink);
+		{
```
Not a must, though I would prefer if the check code was in separate commit from the fix. As these two are isolated changes from each other and one can exist without the other.

With that the authorship will also be less broader than its now.

jimmyyih replied:
Moved to separate commit.

ashwinstar commented on src/test/regress/sql/resource_queue_lock_acquire.sql:
```diff
@@ -0,0 +1,21 @@
+-- Add test description
```
Did we miss adding this test `resource_queue_lock_acquire.sql`  in schedule file?

jimmyyih replied:
https://github.com/greenplum-db/gpdb/pull/12233#discussion_r660877629:
Yeah sorry, this test shouldn't have been added to the PR. It was part of our attempt to put a fault injector in the ResLockAcquire code to simulate a desired error to trigger the try/catch in ResLockPortal in a way that we would hit the assertion failure in ResCleanUpLock... but that didn't work very well so we adopted the isolation2 test that properly creates a deadlock to go into the ResCleanUpLock code path.

jimmyyih commented:
Cleaned up the PR.  I have split the commit into three to make it easier to give authorship to our community contributors.  I have also added a lot of comments to the new isolation2 test to make it easier to follow along with what it's doing.

The PR pipeline and a 5X dev pipeline ran green on the previous iteration of this PR.  Nothing changed logically in the force-push update to this PR so it should be green on the next PR pipeline and 5X dev pipeline run.

Pending PR approvals, this PR is ready for merge into the 5X_STABLE branch.

ashwinstar reviewed (approved):
LGTM. Thank you!

asimrp commented on src/backend/storage/lmgr/lmgr.c:
```diff
@@ -836,6 +836,11 @@ DescribeLockTag(StringInfo buf, const LOCKTAG *tag)
 							 tag->locktag_field2,
 							 tag->locktag_field1);
 			break;
+		case LOCKTAG_RESOURCE_QUEUE:
```
Thank you!

asimrp commented on src/test/isolation2/sql/resource_queue_deadlock.sql:
```diff
@@ -0,0 +1,53 @@
+0<:
+0: SELECT gp_inject_fault_new('before_auto_stats', 'reset', dbid) FROM gp_segment_configuration WHERE content = -1 AND role = 'p';
+
+-- The SELECT will try to acquire ExclusiveLock on the resource
```
The comments are very helpful, thank you.

asimrp reviewed (commented):
👍
